### PR TITLE
Bad play and drive fixes

### PR DESCRIPTION
**Pull request by guga31bb**

### CODE REVIEW

**Comment by guga31bb** [comment]
To summarize:

* If a kickoff is recovered by the kicking team (whether onside kick or fumbled kickoff), a new drive is started
* `posteam` is still receiving team on kickoffs even if it is muffed. This is somewhat weird but I think swapping it would be too confusing

File: R/helper_add_fixed_drives.R [hunk 1]
```diff
@@ -76,6 +76,19 @@ add_drive_results <- function(d) {
...
+      new_drive = dplyr::case_when(
+        .data$play_type == "kickoff" & (.data$own_kickoff_recovery == 1 | .data$fumble_lost == 1) ~ 1,
+        TRUE ~ .data$new_drive
+        ),
+
+      # if it's a kickoff and the prior play was a safety, it's a new drive
+      new_drive = dplyr::case_when(
+        .data$kickoff_attempt == 1 & dplyr::lag(.data$safety) == 1 ~ 1,
+        TRUE ~ .data$new_drive
+      ),
```
Comment (mrcaseb) [inline R/helper_add_fixed_drives.R:90]: Should we try to catch Timeouts between the safety and the kickoff?

File: R/helper_add_fixed_drives.R [hunk 2]
```diff
@@ -15,7 +15,7 @@ add_drive_results <- function(d) {
       posteam = dplyr::case_when(
         # on kickoffs the kicking team is the defteam but this should be swapped
         # in terms of this function if the kickoff is recovered
-        .data$own_kickoff_recovery == 1 ~ .data$defteam,
+        .data$own_kickoff_recovery == 1 | .data$fumble_lost == 1 ~ .data$defteam,
```
Comment (mrcaseb) [inline R/helper_add_fixed_drives.R:18]: This posteam fix should be added to `add_series_data()` as well

Comment (guga31bb) [reply R/helper_add_fixed_drives.R:18]: I don't understand why that would be necessary- series automatically increments upon a new drive right?

Comment (mrcaseb) [reply R/helper_add_fixed_drives.R:18]: posteam is used for the drive result but only for touchdowns. Not 100% sure if it is necessary or not
https://github.com/mrcaseb/nflfastR/blob/1c81c26e02da5bb73cad01192c4909cffc46f92a/R/helper_add_series_data.R#L53-L54

Comment (guga31bb) [reply R/helper_add_fixed_drives.R:18]: Ah so if a team fumbles on a kickoff and it's returned for a TD, we need to decide of the result should count as "Touchdown" or "Opp touchdown". I would say the 2nd?

Comment (mrcaseb) [reply R/helper_add_fixed_drives.R]: It should match the team that makes the PAT because `fixed_drive` won't change. So it should be `"Touchdown"`?

Comment (guga31bb) [reply R/helper_add_fixed_drives.R]: A game with a fumbled kickoff TD for checking
```
g <- fast_scraper("2016_03_NYJ_KC")

g %>%
  # filter(qtr == 2) %>%
  filter(between(play_id, 1200, 1400)) %>%
  select(
    play_id, posteam, desc, fixed_drive, fixed_drive_result, series, series_result
  )
  play_id posteam desc                                                         fixed_drive fixed_drive_res~ series series_result
    <dbl> <chr>   <chr>                                                              <dbl> <chr>             <dbl> <chr>        
1    1210 KC      (7:47) 11-A.Smith pass short right to 87-T.Kelce to NYJ 4 f~           6 Field goal           18 Field goal   
2    1234 KC      (7:04) (Shotgun) 32-S.Ware right end to NYJ 9 for -5 yards ~           6 Field goal           18 Field goal   
3    1255 KC      (6:21) 11-A.Smith pass incomplete short right to 17-C.Conle~           6 Field goal           18 Field goal   
4    1277 KC      (6:15) 5-C.Santos 27 yard field goal is GOOD, Center-41-J.W~           6 Field goal           18 Field goal   
5    1296 NYJ     5-C.Santos kicks 61 yards from KC 35 to NYJ 4. 89-J.Marshal~           7 Touchdown            19 Touchdown    
6    1330 KC      5-C.Santos extra point is GOOD, Center-41-J.Winchester, Hol~           7 Touchdown            19 Touchdown    
7    1346 NYJ     5-C.Santos kicks 65 yards from KC 35 to NYJ 0. 89-J.Marshal~           8 Field goal           20 First down   
8    1379 NYJ     (5:54) (Shotgun) 14-R.Fitzpatrick pass incomplete deep midd~           8 Field goal           20 First down 
```

File: R/helper_add_fixed_drives.R (hunk 1 above)
Comment (guga31bb) [reply R/helper_add_fixed_drives.R]: Done

**Comment by guga31bb** [comment]
My code for replicating [red zone touchdown %](https://www.teamrankings.com/nfl/stat/red-zone-scoring-pct):
```future::plan("multisession")
s <- nflfastR::build_nflfastR_pbp(nflfastR::fast_scraper_schedules(2020))

s %>%
  filter(
    down <= 4
  ) %>%
  group_by(game_id, fixed_drive) %>%
  mutate(
    rz = if_else(yardline_100 <= 19, 1, 0),
    drive_rz = max(rz, na.rm = T)
  ) %>%
  filter(drive_rz == 1) %>%
  summarise(
    drive_result = dplyr::first(fixed_drive_result),
    posteam = dplyr::first(posteam)
  ) %>%
  mutate(drive_td = if_else(drive_result == "Touchdown", 1, 0)) %>%
  group_by(posteam) %>%
  summarize(
    off_rz_td = 100 * mean(drive_td), 
    off_rz_n = n()
  ) %>%
  ungroup() %>%
  arrange(-off_rz_td)

   posteam off_rz_td off_rz_n
   <chr>       <dbl>    <int>
 1 GB           76.8       69
 2 TEN          74.2       66
 3 SEA          73.7       57
 4 CLE          73.3       60
 5 NO           71.6       74
 6 MIN          71.2       59
 7 SF           67.3       55
 8 PIT          66.1       56
 9 DET          66.0       53
10 ARI          65.5       58
# ... with 22 more rows
```

**Pull request merged**